bigtable/google/cloud/bigtable/row_data.py

```diff
@@ -234,25 +239,42 @@ def consume_all(self, max_loops=None):
             self.rows[row.row_key] = row
 
 
+def _retry_read_rows_exception(exc):
+    if isinstance(exc, grpc.RpcError):
+        exc = exceptions.from_grpc_error(exc)
+    return isinstance(exc, (exceptions.ServiceUnavailable,
+                            exceptions.DeadlineExceeded))
+
+
 class YieldRowsData(object):
     """Convenience wrapper for consuming a ``ReadRows`` streaming response.
 
     :type response_iterator: :class:`~google.cloud.exceptions.GrpcRendezvous`
     :param response_iterator: A streaming iterator returned from a
                               ``ReadRows`` request.
+
+    :type request_pb: :class:`data_messages_v2_pb2.ReadRowsRequest`
+    :param request_pb: (Optional) The ``ReadRowsRequest`` protobuf for
+                    the iterator. If present, used to enable retry by
+                    setting a new start_key for the request equal to the
+                    last_scanned_row_key. A new iterator is created, allowing
+                    the scan to continue from the point just beyond the last
+                    successfully read row, identified by last_scanned_row_key.
+
+    :type client: :class:`Client <google.cloud.bigtable.client.Client>`
+    :param client: (Optional) The client that owns the instance. Used for
+                    retry to create a new iterator.
     """
 
     START = 'Start'  # No responses yet processed.
     NEW_ROW = 'New row'  # No cells yet complete for row
     ROW_IN_PROGRESS = 'Row in progress'  # Some cells complete for row
     CELL_IN_PROGRESS = 'Cell in progress'  # Incomplete cell for row
 
-    def __init__(self, response_iterator):
+    def __init__(self, response_iterator, request_pb=None, client=None):
```

Comment:
A couple of things here:

1. Can you just call `request_pb` `request`. It's *not* a protobuf, it's a message, so if you're going to postfix it should be `request_msg`, but there's no need for all that so just `request` will do.
2. Instead of passing the client, just pass in the bound method to execute the request. Call it `method` and pass it in like `YieldRowsData(..., method=client._data_stub.ReadRows)`. This will make the class more flexible and easier to test.
3. It seems you're going out of your way not to break this classes interface and it's actually making things more complex here. This client is not GA, could we consider breaking this so the interface is simply `YieldRowsData(method, request)`? That way, this class would require far less branches to accomplish its task.

Reply:
OK, agreed that this would clear up the interfaces. Will need to publish release notes to make the new interface changes clear to users.  Starting the changes, which will require updates to the unit test cases and will also alter PartialRowsData to be consistent with the new interface.  This will provide the additional benefit of properly supporting retry in PartialRowsData.

Reply:
Latest commits use the suggestions above and the code is now consistent for the cases that used an iterator as a parameter previously. In the Table class, read_row, read_rows and yield_rows all use the suggested change to pass the ReadRows method and ReadRowsRequest as input parameters to the classes implementing the requests.